force-app/main/default/classes/ClassesObjectsMethods.cls

```diff
@@ -212,10 +236,16 @@ public with sharing class ClassesObjectsMethods {
      * @return The a positive number years between the two dates, or null if either date is null.
      */
     public static Integer calculateYearDifference(Date d1, Date d2) {
-
-        // Subtract the difference in years
-
-        return null; // Replace null with the variable you used to store the result
+  
+    if (d1 == null || d2 == null){
+        return null;
+    // Subtract the difference in years
+    } else {
+        Integer d1Year = Datetime.newInstance(d1.year(), 1, 1, 0, 0, 0).yearGmt();
+        Integer d2Year = Datetime.newInstance(d2.year(), 1, 1, 0, 0, 0).yearGmt(); 
+        return Math.abs(d1Year - d2Year);
```

Comment:
Totally correct. But you  are making your life way harder than it needs to be.  Your method params are both Dates just call .year() on them 

```
 public static Integer calculateYearDifference(Date d1, Date d2) {
        if (d1 == null || d2 == null) {
            return null;
        }
        return Math.abs(d1.year() - d2.year());
    }
    ```